devito/passes/iet/misc.py

```diff
@@ -77,47 +90,68 @@ def relax_incr_dimensions(iet, **kwargs):
-            for i, (m, M, b) in zip(outer, p):
-                dynamic_args_mapper[i.symbolic_min] = m
-                dynamic_args_mapper[i.symbolic_max] = M
-                dynamic_args_mapper[i.step] = b
-                for j in inner:
-                    if j.dim.root is i.dim.root and not is_integer(j.step):
-                        value = j.step if b is i.step else b
-                        dynamic_args_mapper[j.step] = (value,)
-            calls.append(efunc.make_call(dynamic_args_mapper))
-
-        mapper[root] = List(body=calls)
-
-    iet = Transformer(mapper).visit(iet)
-
-    return iet, {'efuncs': efuncs}
+        # Get symbolic_max out of each outer dimension
+        roots_max = {i.dim.root: i.symbolic_max for i in outer}
+
+        # A dictionary to map maximum of parent dimensions
+        # useful for hierarchical blocking
+        proc_parents_max = {}
+
+        # Process inner iterations
+        for n, i in enumerate(inner):
+            assert i.direction is Forward
+            if (i.dim.parent in proc_parents_max and
+               i.symbolic_size == i.dim.parent.step):
+                # Special case: A parent dimension may have already been processed
+                # as a member of 'inner' iterations. In this case we can use parent's
+                # Max
+                # Usually encountered in hierarchical blocking (BLOCKLEVELS > 1)
+                it_max = proc_parents_max[i.dim.parent]
+            else:
+                # Most of the cases pass though this code:
+                # Candidates for upper bound calculation are:
+                # Candidate 1: symbolic_max of current iteration
+                # e.g.
+                # i.symbolic_max = x0_blk0 + x0_blk0_size
+                symbolic_max = i.symbolic_max
+
+                # Candidate 2: The domain max. Usualy it is the max of parent/root
+                # dimension.
+                # e.g. x_M
+                # This may not always be true as the symbolic_size of an Iteration may
+                # exceed the size of a parent's block size (e.g. after CIRE passes)
+                # e.g.
+                # i.dim.parent.step = x0_blk1_size
+                # i.symbolic_size = x0_blk1_size + 4
+
+                # For this case, proper margin should be allowed
+                # in order not to drop iterations. So Candidate 2 is the maximum of the
+                # root's max and the current iteration's required max
+                # and instead of `x_M` we may need `x_M + 1` or `x_M + 2`
+                root_max = roots_max[i.dim.root]
+
+                lb_margin = i.symbolic_min - i.dim.symbolic_min
+                size_margin = i.symbolic_size - i.dim.parent.step
+                upper_margin = i.dim.parent.symbolic_max + size_margin + lb_margin
+
+                # Domain max candidate
+                # e.g. domain_max = Max(x_M + 1, x_M)
+                domain_max = Max(upper_margin, root_max)
```

Comment:
so about my comment above

> when isn't the "Iteration's required max" larger than the root's max?

what I meant, IOW, is: why isn't here just `domain_max = upper_margin`

also, nitpicking: I think you could do a better job at choosing variable names, but It might be just me...

Reply:
You are right, need to think of better names

Reply:
https://github.com/devitocodes/devito/pull/1673#discussion_r649293215

Reply:
Fixed